[cache] Infinite loop trying to initialize Cache configuration when deserialization fails

In our case, there may be an application that's joining Hazelcast cluster that doesn't have all classes available to the JCache configuration (i.e. key type and value type)
In this case, trying to deserialize CacheConfig causes ClassNotFoundException.

The expected behavior is that the cache configuration fails, but what's happening that the other node re-transmits cache configuration infinitely, resulting in an infinite loop:

(the exception keeps repeating infinitely with the same stack trace)
```
[2017-06-11T20:00:58.700-0500] [] [INFO] [] [com.hazelcast.system] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1497229258700] [levelValue: 800] [100.74.17.133]:5901 [development] [3.9-SNAPSHOT] Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.

[2017-06-11T20:00:58.700-0500] [] [INFO] [] [com.hazelcast.system] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1497229258700] [levelValue: 800] [100.74.17.133]:5901 [development] [3.9-SNAPSHOT] Configured Hazelcast Serialization version: 1

[2017-06-11T20:00:59.109-0500] [] [INFO] [] [com.hazelcast.spi.impl.operationservice.impl.BackpressureRegulator] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1497229259109] [levelValue: 800] [100.74.17.133]:5901 [development] [3.9-SNAPSHOT] Backpressure is disabled

[2017-06-11T20:00:59.955-0500] [] [INFO] [] [com.hazelcast.instance.Node] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1497229259955] [levelValue: 800] [100.74.17.133]:5901 [development] [3.9-SNAPSHOT] Creating MulticastJoiner

[2017-06-11T20:01:00.178-0500] [] [INFO] [] [com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1497229260178] [levelValue: 800] [100.74.17.133]:5901 [development] [3.9-SNAPSHOT] Starting 4 partition threads

[2017-06-11T20:01:00.179-0500] [] [INFO] [] [com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1497229260179] [levelValue: 800] [100.74.17.133]:5901 [development] [3.9-SNAPSHOT] Starting 3 generic threads (1 dedicated for priority tasks)

[2017-06-11T20:01:00.190-0500] [] [INFO] [] [com.hazelcast.core.LifecycleService] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1497229260190] [levelValue: 800] [100.74.17.133]:5901 [development] [3.9-SNAPSHOT] [100.74.17.133]:5901 is STARTING

[2017-06-11T20:01:00.340-0500] [] [INFO] [] [com.hazelcast.internal.cluster.impl.MulticastJoiner] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1497229260340] [levelValue: 800] [100.74.17.133]:5901 [development] [3.9-SNAPSHOT] Trying to join to discovered node: [100.74.17.133]:5900

[2017-06-11T20:01:00.451-0500] [] [INFO] [] [com.hazelcast.nio.tcp.TcpIpConnector] [tid: _ThreadID=73 _ThreadName=hz._hzInstance_1_development.cached.thread-3] [timeMillis: 1497229260451] [levelValue: 800] [100.74.17.133]:5901 [development] [3.9-SNAPSHOT] Connecting to /100.74.17.133:5900, timeout: 0, bind-any: true

[2017-06-11T20:01:00.455-0500] [] [INFO] [] [com.hazelcast.nio.tcp.TcpIpConnectionManager] [tid: _ThreadID=73 _ThreadName=hz._hzInstance_1_development.cached.thread-3] [timeMillis: 1497229260455] [levelValue: 800] [100.74.17.133]:5901 [development] [3.9-SNAPSHOT] Established socket connection between /100.74.17.133:51741 and /100.74.17.133:5900

[2017-06-11T20:01:06.414-0500] [] [SEVERE] [] [com.hazelcast.config.Config] [tid: _ThreadID=57 _ThreadName=hz._hzInstance_1_development.generic-operation.thread-0] [timeMillis: 1497229266414] [levelValue: 1000] [[
com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.ClassNotFoundException: bg.jug.guestbook.entities.User
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$ClassSerializer.read(JavaDefaultSerializers.java:181)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$ClassSerializer.read(JavaDefaultSerializers.java:169)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.readObject(AbstractSerializationService.java:265)
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:601)
	at com.hazelcast.config.CacheConfig.readData(CacheConfig.java:600)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.readInternal(DataSerializableSerializer.java:158)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:105)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:50)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.readObject(AbstractSerializationService.java:265)
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:601)
	at com.hazelcast.cache.impl.operation.PostJoinCacheOperation.readInternal(PostJoinCacheOperation.java:88)
	at com.hazelcast.spi.Operation.readData(Operation.java:594)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.readInternal(DataSerializableSerializer.java:158)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:105)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:50)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.readObject(AbstractSerializationService.java:265)
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:601)
	at com.hazelcast.internal.cluster.impl.operations.PostJoinOp.readInternal(PostJoinOp.java:124)
	at com.hazelcast.spi.Operation.readData(Operation.java:594)
	at com.hazelcast.internal.cluster.impl.operations.FinalizeJoinOp.readInternalImpl(FinalizeJoinOp.java:156)
	at com.hazelcast.internal.cluster.impl.operations.VersionedClusterOperation.readInternal(VersionedClusterOperation.java:59)
	at com.hazelcast.spi.Operation.readData(Operation.java:594)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.readInternal(DataSerializableSerializer.java:158)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:105)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:50)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:183)
	at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:351)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:394)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:115)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
Caused by: java.lang.ClassNotFoundException: bg.jug.guestbook.entities.User
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at com.sun.enterprise.loader.CurrentBeforeParentClassLoader.loadClass(CurrentBeforeParentClassLoader.java:83)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at com.hazelcast.nio.ClassLoaderUtil.tryLoadClass(ClassLoaderUtil.java:173)
	at com.hazelcast.nio.ClassLoaderUtil.loadClass(ClassLoaderUtil.java:147)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$ClassSerializer.read(JavaDefaultSerializers.java:179)
	... 33 more
]]
```

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/FinalizeJoinOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/FinalizeJoinOp.java
@@ -18,6 +18,7 @@ package com.hazelcast.internal.cluster.impl.operations;
 
 import com.hazelcast.cluster.ClusterState;
 import com.hazelcast.core.Member;
+import com.hazelcast.instance.Node;
 import com.hazelcast.internal.cluster.impl.ClusterDataSerializerHook;
 import com.hazelcast.internal.cluster.impl.ClusterServiceImpl;
 import com.hazelcast.internal.cluster.impl.MembersView;
@@ -29,6 +30,7 @@ import com.hazelcast.spi.Operation;
 import com.hazelcast.spi.OperationAccessor;
 import com.hazelcast.spi.OperationService;
 import com.hazelcast.spi.impl.NodeEngineImpl;
+import com.hazelcast.util.ExceptionUtil;
 import com.hazelcast.version.Version;
 
 import java.io.IOException;
@@ -55,6 +57,7 @@ public class FinalizeJoinOp extends MembersUpdateOp {
     private Version clusterVersion;
 
     private transient boolean finalized;
+    private transient Exception deserializationFailure;
 
     public FinalizeJoinOp() {
     }
@@ -83,6 +86,8 @@ public class FinalizeJoinOp extends MembersUpdateOp {
         Address callerAddress = getConnectionEndpointOrThisAddress();
         String callerUuid = getCallerUuid();
 
+        checkDeserializationFailure(clusterService);
+
         preparePostOp(preJoinOp);
         finalized = clusterService.finalizeJoin(getMembersView(), callerAddress, callerUuid,
                 clusterId, clusterState, clusterVersion, clusterStartTime, masterTime, preJoinOp);
@@ -92,6 +97,15 @@ public class FinalizeJoinOp extends MembersUpdateOp {
         }
 
         processPartitionState();
+    }
+
+    private void checkDeserializationFailure(ClusterServiceImpl clusterService) {
+        if (deserializationFailure != null) {
+            getLogger().severe("Node could not join cluster.", deserializationFailure);
+            Node node = clusterService.getNodeEngine().getNode();
+            node.shutdown(true);
+            throw ExceptionUtil.rethrow(deserializationFailure);
+        }
     }
 
     @Override
@@ -167,15 +181,26 @@ public class FinalizeJoinOp extends MembersUpdateOp {
         boolean hasPostJoinOp = in.readBoolean();
         if (hasPostJoinOp) {
             postJoinOp = new OnJoinOp();
-            postJoinOp.readData(in);
+            try {
+                postJoinOp.readData(in);
+            } catch (Exception e) {
+                deserializationFailure = e;
+                // return immediately, do not attempt to read further
+                return;
+            }
         }
         clusterId = in.readUTF();
         clusterStartTime = in.readLong();
         String stateName = in.readUTF();
         clusterState = ClusterState.valueOf(stateName);
         clusterVersion = in.readObject();
+
         if (clusterVersion.isGreaterOrEqual(V3_9)) {
-            preJoinOp = in.readObject();
+            try {
+                preJoinOp = in.readObject();
+            } catch (Exception e) {
+                deserializationFailure = e;
+            }
         }
     }
 

--- a/hazelcast/src/test/java/classloading/domain/Person.java
+++ b/hazelcast/src/test/java/classloading/domain/Person.java
@@ -1,0 +1,26 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package classloading.domain;
+
+import java.io.Serializable;
+
+/**
+ * Sample domain class defined outside com.hazelcast package for classloader-related tests
+ */
+public class Person implements Serializable {
+
+}

--- a/hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/MembershipUpdateTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/MembershipUpdateTest.java
@@ -16,7 +16,10 @@
 
 package com.hazelcast.internal.cluster.impl;
 
+import classloading.domain.Person;
+import com.hazelcast.cache.impl.HazelcastServerCachingProvider;
 import com.hazelcast.cluster.Joiner;
+import com.hazelcast.config.CacheConfig;
 import com.hazelcast.config.Config;
 import com.hazelcast.config.ServiceConfig;
 import com.hazelcast.core.HazelcastInstance;
@@ -43,13 +46,16 @@ import com.hazelcast.test.RequireAssertEnabled;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.util.FilteringClassLoader;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import javax.cache.spi.CachingProvider;
 import java.nio.channels.ServerSocketChannel;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.CountDownLatch;
@@ -667,6 +673,28 @@ public class MembershipUpdateTest extends HazelcastTestSupport {
         sleepSeconds(5);
         sendOpsFromMaster.cancel(true);
         assertFalse(service.otherOpExecutedBeforePreJoin.get());
+    }
+
+    @Test
+    public void shouldShutdown_whenExceptionDeserializing_prePostJoinOps() {
+        // create a HazelcastInstance with a CacheConfig referring to a Class not resolvable on the joining member
+        HazelcastInstance hz1 = factory.newHazelcastInstance();
+        CachingProvider cachingProvider = HazelcastServerCachingProvider.createCachingProvider(hz1);
+        cachingProvider.getCacheManager().createCache("test", new CacheConfig<String, Person>()
+                .setTypes(String.class, Person.class).setManagementEnabled(true));
+
+        // joining member cannot resolve Person class
+        List<String> excludes = Arrays.asList("classloading");
+        ClassLoader classLoader = new FilteringClassLoader(excludes, null);
+        Config config = new Config();
+        config.setClassLoader(classLoader);
+        try {
+            factory.newHazelcastInstance(config);
+            fail("Second HazelcastInstance should not have started");
+        } catch (IllegalStateException e) {
+            // expected
+        }
+        assertClusterSize(1, hz1);
     }
 
     private Config getConfigWithPreJoinAwareService(PreJoinAwareService service) {

